File: ProblemSolving/450DSA/JavaScript/src/arrays/CyclicallyRotateAnArrayByOne.js
/**
 * @author Anirudh Sharma
 *
 * Given an array, cyclically rotate an array by one.
 */
const rotateArray = (a) => {
    // Base condition
    if (a === undefined || a.length <= 1) {
        return a;
    }
    // Save the last element of the array for future
    const lastElement = a[a.length - 1];
    // Index pointing at the end of the array
    let i = a.length - 2;
    while (i >= 0) {
        // Copy the current element to its right element
        a[i + 1] = a[i];
        i--;
    }
    // Put the last element at first position (0th index)
    a[0] = lastElement;
    return a;
};

console.log(rotateArray([1, 2, 3, 4, 5]));
console.log(rotateArray([1, 2]));
console.log(rotateArray([1]));
console.log(rotateArray([]));

File: ProblemSolving/450DSA/Python/src/arrays/CyclicallyRotateAnArrayByOne.py
"""
@author Anirudh Sharma

Given an array, cyclically rotate an array by one.
"""


def rotateArray(a):
    # Base condition
    if a == None or len(a) <= 1:
        return a
    # Save the last element of the array for future
    lastElement = a[len(a) - 1]
    # Index pointing at the end of the array
    i = len(a) - 2
    while i >= 0:
        a[i + 1] = a[i]
        i -= 1
    # Put last element at the start (0th index)
    a[0] = lastElement
    return a


if __name__ == "__main__":
    print(rotateArray([1, 2, 3, 4, 5]))
    print(rotateArray([1, 2]))
    print(rotateArray([1]))
    print(rotateArray([]))
